File: src/catprinter/ble.py
import asyncio
import contextlib
import uuid
from typing import Optional

from bleak import BleakClient, BleakScanner
from bleak.backends.scanner import AdvertisementData
from bleak.backends.device import BLEDevice
from catprinter import logger

# For some reason, bleak reports the 0xaf30 service on my macOS, while it reports
# 0xae30 (which I believe is correct) on my Raspberry Pi. This hacky workaround
# should cover both cases.

POSSIBLE_SERVICE_UUIDS = [
    '0000ae30-0000-1000-8000-00805f9b34fb',
    '0000af30-0000-1000-8000-00805f9b34fb',
]

TX_CHARACTERISTIC_UUID = '0000ae01-0000-1000-8000-00805f9b34fb'

SCAN_TIMEOUT_S = 10

# Wait time after sending each chunk of data through BLE.
WAIT_AFTER_EACH_CHUNK_S = 0.02

# This is a hacky solution so we don't terminate the BLE connection to the printer
# while it's still printing. A better solution is to subscribe to the RX characteristic
# and listen for printer events, so we know exactly when the printing is finished.
WAIT_AFTER_DATA_SENT_S = 30


async def scan(name: Optional[str], timeout: int):
    autodiscover = (not name)
    if autodiscover:
        logger.info('⏳ Trying to auto-discover a printer...')
    else:
        logger.info(f'⏳ Looking for a BLE device named {name}...')

    def filter_fn(device: BLEDevice, adv_data: AdvertisementData):
        if autodiscover:
            return any(uuid in adv_data.service_uuids
                       for uuid in POSSIBLE_SERVICE_UUIDS)
        else:
            return device.name == name

    device = await BleakScanner.find_device_by_filter(
        filter_fn, timeout=timeout,
    )
    if device is None:
        raise RuntimeError('Unable to find printer, make sure it is turned on and in range')
    logger.info(f'✅ Got it. Address: {device}')
    return device


def chunkify(data, chunk_size):
    return (
        data[i: i + chunk_size] for i in range(0, len(data), chunk_size)
    )


async def get_device_address(device: Optional[str]):
    # See if we were passed a string that smells like an UUID or MAC address.
    if device:
        with contextlib.suppress(ValueError):
            return str(uuid.UUID(device))
        if device.count(':') == 5 and device.replace(':', '').isalnum():
            return device

    return await scan(device, timeout=SCAN_TIMEOUT_S)


async def run_ble(data, device: Optional[str]):
    try:
        address = await get_device_address(device)
    except RuntimeError as e:
        logger.error(f'🛑 {e}')
        return
    logger.info(f'⏳ Connecting to {address}...')
    async with BleakClient(address) as client:
        logger.info(
            f'✅ Connected: {client.is_connected}; MTU: {client.mtu_size}')
        chunk_size = client.mtu_size - 3
        logger.info(
            f'⏳ Sending {len(data)} bytes of data in chunks of {chunk_size} bytes...')
        logger.info(
            f'⏳ Sending {(data)} ')
        # for i, chunk in enumerate(chunkify(data, chunk_size)):
        #     logger.info(
        #         f'⏳ Sending {chunk}')
        #     await client.write_gatt_char(TX_CHARACTERISTIC_UUID, chunk)
        #     await asyncio.sleep(WAIT_AFTER_EACH_CHUNK_S)
            # logger.info(f'{chunk}')
        logger.info(f'✅ Done. Waiting {WAIT_AFTER_DATA_SENT_S}s before disconnecting...')
        await asyncio.sleep(WAIT_AFTER_DATA_SENT_S)

File: src/node/ble.ts
import * as ble from 'node-ble'
import { CatImage} from "./image";
import debug_lib, {Debugger} from 'debug';
import * as cmd from './commands';

const POSSIBLE_SERVICE_UUIDS = [
    '0000ae30-0000-1000-8000-00805f9b34fb',
    '0000af30-0000-1000-8000-00805f9b34fb',
]

const PRINT_CHARACTERISTIC = "0000ae01-0000-1000-8000-00805f9b34fb"
const NOTIFY_CHARACTERISTIC = "0000ae02-0000-1000-8000-00805f9b34fb"

// Wait time after sending each chunk of data through BLE.
const WAIT_AFTER_EACH_CHUNK_MS = 20

// This is a hacky solution so we don't terminate the BLE connection to the printer
// while it's still printing. A better solution is to subscribe to the RX characteristic
// and listen for printer events, so we know exactly when the printing is finished.
const WAIT_AFTER_DATA_SENT_MS = 20000

const {bluetooth, destroy} = ble.createBluetooth()

const sleep = (ms: number) => new Promise(accept => setTimeout(accept, ms));

export class CatPrinter {
    private debugger: Debugger
    public device: ble.Device | undefined
    public print_characteristic: ble.GattCharacteristic | undefined
    public notify_characteristic: ble.GattCharacteristic | undefined

    private constructor() {
        this.debugger = debug_lib('cat')
    }

    public async scan(address?: string, timeout_ms?: number): Promise<void> {
        let timeout: number
        let autodiscover: boolean
        let notify_characteristic: ble.GattCharacteristic
        let printer_characteristic: ble.GattCharacteristic
        timeout_ms ? timeout = timeout_ms : timeout = 10000
        address? autodiscover = true : autodiscover = false
        const adapter = await bluetooth.defaultAdapter()

        return new Promise<void>( async (resolve, reject) => {
            async function on_timeout_reached() {
                await adapter.stopDiscovery()
                destroy()
                reject('Unable to find printer, make sure it is turned on and in range')
            }

            const t = setTimeout(on_timeout_reached, timeout)
            t

            if (! await adapter.isDiscovering()) {
                await adapter.startDiscovery()
            }

            if (address) {
                this.debugger(`⏳ Looking for a BLE device with address: ${address}...`)
                try {
                    const device = await adapter.waitDevice(address)
                    this.debugger('Discovered Bluetooth printer')
                    clearTimeout(t)

                    await device.connect()
                    this.debugger('Connected to device')
                    
                    const gattServer = await device.gatt()
                    const services = await gattServer.services()

                    for (let service of services) {
                        let srv = await gattServer.getPrimaryService(service)
                        let chars = await srv.characteristics()

                        for (let char of chars) {
                            const print_char = await srv.getCharacteristic(char)
                            if (char == NOTIFY_CHARACTERISTIC) {
                                notify_characteristic = print_char
                            }
                            if (char == PRINT_CHARACTERISTIC) {
                                printer_characteristic = print_char
                            }
                        }
                    }
                        
                    if (((printer_characteristic != undefined) && (notify_characteristic != undefined))) {
                        this.device = device
                        this.print_characteristic = printer_characteristic
                        this.notify_characteristic = notify_characteristic
                        this.debugger(`✅ Done. Connected to ${await this.device.getAddress()}`)
                        resolve()
                    }
                } catch (err) {
                    this.debugger('Catched error: ', err)
                    destroy()
                    reject(err)
                }
            } else {
                this.debugger(`⏳ Trying to auto-discover a printer...`)
                //TODO       
            }
        })
    }

    public async sendImage(image_path: string, dark_mode?: boolean): Promise<void> {
        let img: CatImage = await CatImage.loadFromPath(image_path)
        //!
        img.save()
        //!
        let data = await cmd.commandsPrintImg(img, dark_mode=dark_mode)
        await this.write(data)
        return
    }


    // public async sendText(text: string): Promise<void> {
    //     let img: CatImage = await CatImage.drawText(text)
    //     let data = await cmd.commandsPrintImg(img, true)
    //     return await this.write(data)
    // }

    public async disconnect(): Promise<void> {
        await this.device?.disconnect()
        this.debugger(`⏳ Disconnecting from the printer...`)
        destroy()
        return
    }

    private async write(data: number[]): Promise<void> {
        const chunk_size = 20
        this.debugger(`⏳ Sending ${data.length} bytes of data in chunks of ${chunk_size} bytes...`)
        
        for (const chunk of (this.chunkify(data, chunk_size))) {
            this.print_characteristic!.writeValueWithoutResponse(Buffer.from(chunk))
            await sleep(WAIT_AFTER_EACH_CHUNK_MS)
        }
        this.debugger(`✅ Done. Waiting ${WAIT_AFTER_DATA_SENT_MS}s before disconnecting...`)
        await sleep(WAIT_AFTER_DATA_SENT_MS)
        return
    }

    private chunkify(data: number[], chunk_size: number): number[][] {
        const chunks: number[][] = []
        for (let i = 0; i < data.length; i += chunk_size) {
            chunks.push(data.slice(i, i + chunk_size))
        }
        return chunks
    }
}
